lib/connection_pool.rb

```diff
@@ -97,6 +90,27 @@ def shutdown(&block)
 
   private
 
+  def checkout_with_state(options)
+    # Handles cleaning up if we timeout in an ensure block and the conn
+    # is automatically popped from the local stack and not the global one.
+    stack.delete_if do |conn|
```

Comment:
This method is run for every single checkout.  Imagine a pool of 50 connections: we iterate through all 50 every time we check anything out.  I'd rather see a while loop that tests the minimum number of connections until we get a good one.